src/System.Net.HttpListener/src/System/Net/WebSockets/WebSocketProtocolComponent.cs

```diff
@@ -0,0 +1,709 @@
+            internal uint ValueLength;
+        }
+
+        [SecuritySafeCritical]
+        static WebSocketProtocolComponent()
+        {
+            s_WebSocketDllHandle = Interop.mincore.LoadLibraryExW(Interop.Libraries.WebSocket, IntPtr.Zero, 0);
+
+            if (!s_WebSocketDllHandle.IsInvalid)
+            {
+                s_SupportedVersion = GetSupportedVersion();
+
+                s_ServerFakeRequestHeaders = new HttpHeader[]
+                {
+                    new HttpHeader()
+                    {
+                        Name = HttpKnownHeaderNames.Connection,
+                        NameLength = (uint)HttpKnownHeaderNames.Connection.Length,
+                        Value = HttpKnownHeaderNames.Upgrade,
+                        ValueLength = (uint)HttpKnownHeaderNames.Upgrade.Length
+                    },
+                    new HttpHeader()
+                    {
+                        Name = HttpKnownHeaderNames.Upgrade,
+                        NameLength = (uint)HttpKnownHeaderNames.Upgrade.Length,
+                        Value = WebSocketHelpers.WebSocketUpgradeToken,
+                        ValueLength = (uint)WebSocketHelpers.WebSocketUpgradeToken.Length
+                    },
+                    new HttpHeader()
+                    {
+                        Name = HttpKnownHeaderNames.Host,
+                        NameLength = (uint)HttpKnownHeaderNames.Host.Length,
+                        Value = string.Empty,
+                        ValueLength = 0
+                    },
+                    new HttpHeader()
+                    {
+                        Name = HttpKnownHeaderNames.SecWebSocketVersion,
+                        NameLength = (uint)HttpKnownHeaderNames.SecWebSocketVersion.Length,
+                        Value = s_SupportedVersion,
+                        ValueLength = (uint)s_SupportedVersion.Length
+                    },
+                    new HttpHeader()
+                    {
+                        Name = HttpKnownHeaderNames.SecWebSocketKey,
+                        NameLength = (uint)HttpKnownHeaderNames.SecWebSocketKey.Length,
+                        Value = s_DummyWebsocketKeyBase64,
+                        ValueLength = (uint)s_DummyWebsocketKeyBase64.Length
+                    }
+                };
+            }
+        }
+
+        internal static string SupportedVersion
+        {
+            get
+            {
+                if (s_WebSocketDllHandle.IsInvalid)
+                {
+                    WebSocketHelpers.ThrowPlatformNotSupportedException_WSPC();
+                }
+
+                return s_SupportedVersion;
+            }
+        }
+
+        internal static bool IsSupported
+        {
+            get
+            {
+                return !s_WebSocketDllHandle.IsInvalid;
+            }
+        }
+
+        [DllImport(Interop.Libraries.WebSocket, EntryPoint = "WebSocketCreateClientHandle", ExactSpelling = true)]
```

Comment:
This needs some extra work: the Interop layer needs to be properly split into separate files, etc.